agent/src/main/java/com/thoughtworks/go/agent/service/AgentUpgradeService.java

```diff
@@ -91,6 +97,11 @@ void checkForUpgrade(String agentMd5, String launcherMd5, String agentPluginsMd5
         }
     }
 
+    private void updateExtraProperties(Header extraPropertiesHeader) {
+        final Map<String, String> extraProperties = parseExtraProperties(extraPropertiesHeader);
+        extraProperties.forEach(System::setProperty);
```

Comment:
All of this looks good to me. Can merge this as is.

If any of the properties change server side, can/should we force a restart? Some JVM specific system properties cannot be changed at runtime (encoding, ssl debug, networking).

Reply:
@ketan I considered that. I thought it would be too disruptive since it will restart all agents immediately if that happens. I'd prefer to not do it unless we find a need for it.